### PR TITLE
[python, ruby] Add Director guard file with C++ 11 std::mutex

### PR DESCRIPTION
**Pull request by erezgeva**
Follow #2861

Add a common file for python and Ruby.
Remove unused parts from ocaml director.swg.
Add C++ 11 for newer compilers and improve platform independent.

### CODE REVIEW

**Comment by wsfulton** [comment]
ping @erezgeva, can you address these small issues?

**Comment by erezgeva** [comment]
> ping @erezgeva, can you address these small issues?

This pull request impleament #2861.

Do you mean to #2209?

I think my patch solve it.
On most Windows it will use C++ 11 std::mutex and with old Windows the CRITICAL_SECTION is encapsulate in a class.
But I do not have a Windows machine to test it.
I only use the GitHub actions with windows here.

Erez

**Review by wsfulton — CHANGES_REQUESTED** [review]

**Comment by wsfulton** [comment]
> ping @erezgeva, can you address these small issues?

I was referring to my review comments, but they were still pending so you wouldn't have seen them! I've just posted them now, so sorry for the confusion.

File: Lib/director_guard.swg [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+/*
+ * C++ 11
+ * std::mutex        https://en.cppreference.com/w/cpp/thread/mutex
+ * std::unique_lock  https://en.cppreference.com/w/cpp/thread/unique_lock
+ */
+#include <mutex>
+#ifdef SWIG_HAVE_MUTEX /* Use Language defined Mutex class */
+#define SWIG_GUARD(_mutex) std::unique_lock<Mutex> _guard(_mutex)
+#define SWIG_GUARD_DEF(_cls, _mutex) Mutex _cls::_mutex
+#define SWIG_GUARD_DEC(_mutex) static Mutex _mutex
```
[the remark this answers is not in the document]
Comment (erezgeva) [reply Lib/director_guard.swg:25]: Done

File: Source/Modules/python.cxx [hunk 2]
```diff
@@ -808,6 +808,9 @@ class PYTHON:public Language {
     if (Swig_directors_enabled()) {
       // Insert director runtime into the f_runtime file (make it occur before %header section)
       Swig_insert_file("director_common.swg", f_runtime);
+      Swig_insert_file("director_common.swg", f_runtime);
```
[the remark this answers is not in the document]
Comment (erezgeva) [reply Source/Modules/python.cxx:811]: Fixed

File: Lib/director_guard.swg [hunk 3]
```diff
@@ -0,0 +1,94 @@
+/* -----------------------------------------------------------------------------
+ * director_guard.swg
+ *
+ * Generic Mutex implementation
```
[the remark this answers is not in the document]
Comment (erezgeva) [reply Lib/director_guard.swg:4]: I try to improve description.

**Comment by wsfulton** [comment]
I've added some docs in 5cfe3e7b3c656d395e530a445b782e4f4c25fbfd for my understanding of this. Please check. The threading was turned off by default in both Python and Ruby and now we have to defined `SWIG_THREADS` to turn the mutex locks on. We should have a test that tests `SWIG_THREADS`. Perhaps add a new test that includes an existing director test and test it on say just c++11:
```swig
%begin %{
#if __cplusplus >= 201103L
#define SWIG_THREADS
#endif
%}
```
But it would be good to have a testcase that replicates the original bug, but maybe that isn't easy to construct.

**Comment by erezgeva** [comment]
> ```c++
> SWIG_THREADS
> ```

I tried to leave thing as their were.

In Ruby the mutex code is not used at the moment.
In python it is done with:
swig command line: 
`     -threads        - Add thread support for all the interface`
Source/Modules/python.cxx:
`    } else if (threads) {`
`      Printf(f_runtime, "#define SWIG_PYTHON_THREADS\n");`
Or using the `%thread` in the interface file.

A test is a good idea.
I'll try write one for python.
We could see that the C++ code define the Mutex.
May be print the class code result of `gcc -E` in the log ~20-30 lines?

As for the bug.
I do not have a Windows machine.
I think it is better to ask the person who found the bug and ask him to test.

**Comment by wsfulton** [comment]
> A test is a good idea.
> I'll try write one for python.
> We could see that the C++ code define the Mutex.
> May be print the class code result of gcc -E in the log ~20-30 lines?

No need to add `gcc -E` output into the test, the test-suite doesn't handle that, however, checking this manually is a good idea. 

> As for the bug.
> I do not have a Windows machine.
> I think it is better to ask the person who report the bug and ask him to test.

Yes, that might be all we can do. I thought you saw a crash too and if you can replicate in a simple **runtime** test, then you can add that to the test-suite.

**Comment by erezgeva** [comment]
> > A test is a good idea.
> > I'll try write one for python.
> > We could see that the C++ code define the Mutex.
> > May be print the class code result of gcc -E in the log ~20-30 lines?
> 
> No need to add `gcc -E` output into the test, the test-suite doesn't handle that, however, checking this manually is a good idea.
> 
> > As for the bug.
> > I do not have a Windows machine.
> > I think it is better to ask the person who report the bug and ask him to test.
> 
> Yes, that might be all we can do. You saw a problem too ("corrupted double-linked list" mentioned in #2861) and if you can replicate in a simple **runtime** test, then you can add that to the test-suite. With `SWIG_THREADS` defined in the test, it'll get tested on Windows (for Python) in the test-suite on Appveyor.

I add python also in #2813 win_ci.yml.

Regarding the "corrupted double-linked list", I saw. The problem was in my code, not SWIG.
I have a run condition between two destruction not created by SWIG,.
The run condition happens with my Ruby wrapper, and I was not sure.

I'll try write a test in python for this code.

**Comment by wsfulton** [comment]
> I'll try write a test in python for this code.

Thanks for doing this (#2949).

**Pull request merged**